### PR TITLE
Non_empty_list type

### PR DESCRIPTION
**Pull request by bkase**
I keep finding myself wanting this and getting upset that we have to have
runtime assertions, so I just went ahead and made this.

These functions all delegate to List so it should be pretty easy to reason that this code is correct.

### CODE REVIEW

**Comment by psteckler** [comment]
Maybe useful to have a `singleton` function.

**Comment by bkase** [comment]
> Could you have a tail_exn that preserves the non-emptiness when possible, else raises an exn?

Given that I'm making this to avoid runtime checks, I'll add a `tail_opt`, but require consumers to `Option.value_exn` it if they really want to forgo the compile-time safety.

File: src/lib/non_empty_list/non_empty_list.ml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+(* A non-empty list is a tuple of the head and the rest (as a list) *)
+type 'a t = 'a * 'a list [@@deriving sexp, compare, eq, hash]
+
+let init x xs = (x, xs)
+
+let singleton x = (x, [])
+
+let uncons = Fn.id
+
+let cons (x, xs) x' = (x', x :: xs)
```
Comment (cmr) [inline src/lib/non_empty_list/non_empty_list.ml]: shouldn't the list be the second argument of cons?

Comment (psteckler) [reply src/lib/non_empty_list/non_empty_list.ml]: +1 what @cmr said.

File: src/lib/non_empty_list/non_empty_list.ml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+let head (x, _) = x
+
+let tail (_, xs) = xs
+
+let of_list = function [] -> None | x :: xs -> Some (x, xs)
```
Comment (psteckler) [inline src/lib/non_empty_list/non_empty_list.ml]: should this be `of_list_opt` ?

**Pull request merged**